File: 封本/apps/dto/api/arbitration.py
"""
apps/dto/api/arbitration.py — 仲裁层 API DTO

API 层接收 DTO，内部转换为核心对象后调用 ArbitrationEngine。
AI 只能通过这些 DTO 与仲裁层交互，禁止直接绑定 ArbitrationDecision。
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field, field_validator


# ── 请求 ────────────────────────────────────────────────────

class ArbitrationRunRequest(BaseModel):
    """
    POST /arbitration/run 请求（旧入口：消费 Phase 5 信号）。

    字段设计：
    - 最小可用：symbol + direction + confidence（走通主链路）
    - 扩展字段：regime / regime_confidence（影响 RegimeFilterRule）
    - 扩展字段：fundamental_score（影响 FundamentalVetoRule）
    """

    symbol: str = Field(min_length=1, max_length=16)
    direction: Literal["LONG", "SHORT", "FLAT"] = Field(description="信号方向")
    confidence: float = Field(ge=0.0, le=1.0, description="信号置信度")
    strength: float = Field(default=0.5, ge=0.0, le=1.0)

    # 扩展字段（可选，影响规则链）
    regime: Literal["trending_up", "trending_down", "ranging", "volatile", "unknown"] = (
        Field(default="trending_up")
    )
    regime_confidence: float = Field(default=0.5, ge=0.0, le=1.0)
    fundamental_score: float = Field(
        default=0.5, ge=0.0, le=1.0,
        description="基本面评分，影响 FundamentalVetoRule",
    )
    orderflow_score: float = Field(default=0.5, ge=0.0, le=1.0)
    sentiment_score: float = Field(default=0.5, ge=0.0, le=1.0)
    macro_score: float = Field(default=0.5, ge=0.0, le=1.0)

    timestamp: Optional[datetime] = Field(default=None, description="时间戳，默认当前时间")


class PortfolioArbitrationRequest(BaseModel):
    """
    POST /arbitration/run-portfolio 请求（新入口：消费 Phase 9 策略池）。

    策略信号作为请求体传入，不引用核心 StrategySignalBundle 类型。
    """

    portfolio_id: str = Field(description="组合 ID，如 AAPL-SP")
    symbol: str = Field(description="标的代码，如 AAPL")

    # 策略提案（可多个，模拟 Phase 9 输出）
    proposals: list["StrategyProposalRequest"] = Field(
        min_length=1, description="至少一个策略提案"
    )

    timestamp: Optional[datetime] = Field(default=None)

    @field_validator("portfolio_id")
    @classmethod
    def _strip(cls, v: str) -> str:
        return v.strip()


class StrategyProposalRequest(BaseModel):
    """策略提案（嵌入 DTO，不引用核心 StrategySignalBundle）。"""

    proposal_id: str
    strategy_id: str = Field(description="策略标识，如 trend / mean_reversion")
    aggregate_direction: Literal["LONG", "SHORT", "FLAT"]
    aggregate_strength: float = Field(ge=0.0, le=1.0)
    aggregate_confidence: float = Field(ge=0.0, le=1.0)
    portfolio_weight: float = Field(default=1.0, ge=0.0, le=1.0)

    # 可选：bundle 列表（模拟 StrategySignalBundle 输出）
    bundle_count: int = Field(default=1, ge=1, description="bundle 数量")


# ── 响应 ────────────────────────────────────────────────────

class DecisionRationaleView(BaseModel):
    """决策理由（展示用 view model）。"""

    signal_name: str
    direction: str
    confidence: float
    weight: float
    contribution: float
    rule_adjustments: list[str] = Field(default_factory=list)


class ArbitrationResponse(BaseModel):
    """
    仲裁决策响应（与核心 ArbitrationDecision 完全解耦）。

    AI 通过此 DTO 接收结果，禁止直接访问 ArbitrationDecision 对象。
    """

    ok: bool = True
    decision_id: str = Field(description="决策 ID")
    symbol: str
    bias: str = Field(description="long_bias | short_bias | hold_bias | no_trade")
    confidence: float = Field(ge=0.0, le=1.0)
    signal_count: int = Field(ge=0, description="参与评分的信号数量")
    rules_applied: list[str] = Field(
        default_factory=list,
        description="应用的规则名称列表",
    )
    rationale: list[DecisionRationaleView] = Field(
        default_factory=list,
        description="各信号理由",
    )
    timestamp: datetime
    arbitration_latency_ms: float = Field(ge=0.0)
    source: str = Field(
        default="arbitration",
        description="来源：arbitration（旧入口）| portfolio（策略池入口）",
    )

    model_config = {"str_strip_whitespace": True}


# 修复 forward reference
PortfolioArbitrationRequest.model_rebuild()

File: TradeOS/apps/web_console/assets/views/arbitration.js
function renderDecisionSummary(result, utils, emptyText) {
  if (!result) {
    return `<div class="empty-state compact">${utils.escapeHtml(emptyText)}</div>`;
  }
  return `
    ${utils.renderMetricCards([
      {
        label: "偏置 Bias",
        value: result.bias || "n/a",
        sub: `Decision ${utils.shortText(result.decision_id, 18)}`,
      },
      {
        label: "置信度 Confidence",
        value: utils.formatNumber(result.confidence),
        sub: `信号数量 Signal Count ${result.signal_count ?? "n/a"}`,
      },
      {
        label: "来源 Source",
        value: result.source || "n/a",
        sub: `Latency ${utils.formatNumber(result.arbitration_latency_ms, 1)} ms`,
      },
    ])}
    <div class="mini-card">
      <span class="mini-label">已应用规则 Rules Applied</span>
      <div class="pill-row">${utils.renderPillList(result.rules_applied || [], "success")}</div>
    </div>
  `;
}

export const arbitrationView = {
  id: "arbitration",
  label: "仲裁 Arbitration",
  shortLabel: "决策 Bias",
  initialState: {
    singleForm: {
      symbol: "AAPL",
      direction: "LONG",
      confidence: 0.8,
      strength: 0.6,
      regime: "trending_up",
    },
    portfolioForm: {
      portfolio_id: "AAPL-SP",
      symbol: "AAPL",
      trend_direction: "LONG",
      trend_strength: 0.72,
      trend_confidence: 0.76,
      trend_weight: 0.55,
      mean_direction: "FLAT",
      mean_strength: 0.35,
      mean_confidence: 0.48,
      mean_weight: 0.45,
    },
    singleResult: null,
    portfolioResult: null,
  },
  render(ctx, page) {
    const { utils } = ctx;
    const rationale = page.singleResult?.rationale || [];

    return `
      <section class="page-head">
        <div>
          <span class="section-eyebrow">决策层 Decision Layer</span>
          <h2>仲裁 Arbitration</h2>
          <p>单标的仲裁与组合仲裁都直接调用当前后端。单标的表单只提交后端当前真正接受的 DTO 字段。</p>
        </div>
        <div class="head-actions">
          <span class="endpoint-pill">POST /api/v1/arbitration/run</span>
          <span class="endpoint-pill">POST /api/v1/arbitration/run-portfolio</span>
        </div>
      </section>

      <section class="panel-grid two-up">
        <article class="panel-card">
          <div class="panel-head">
            <div>
              <span class="panel-kicker">单标的 Single Symbol</span>
              <h3>运行仲裁 Run Arbitration</h3>
            </div>
            <span class="helper-copy">提交字段与后端 ArbitrationRunRequest 保持一致：symbol、direction、confidence、strength、regime。</span>
          </div>
          <form data-single-form class="form-grid">
            <label><span>标的 Symbol</span><input name="symbol" value="${utils.escapeHtml(page.singleForm.symbol)}" /></label>
            <label>
              <span>方向 Direction</span>
              <select name="direction">
                ${["LONG", "SHORT", "FLAT"]
                  .map((option) => `<option value="${option}" ${page.singleForm.direction === option ? "selected" : ""}>${option}</option>`)
                  .join("")}
              </select>
            </label>
            <label>
              <span>市场状态 Regime</span>
              <select name="regime">
                ${["trending_up", "trending_down", "ranging", "volatile", "unknown"]
                  .map((option) => `<option value="${option}" ${page.singleForm.regime === option ? "selected" : ""}>${option}</option>`)
                  .join("")}
              </select>
            </label>
            <label><span>置信度 Confidence</span><input name="confidence" type="number" min="0" max="1" step="0.01" value="${utils.escapeHtml(page.singleForm.confidence)}" /></label>
            <label><span>强度 Strength</span><input name="strength" type="number" min="0" max="1" step="0.01" value="${utils.escapeHtml(page.singleForm.strength)}" /></label>
          </form>
          <div class="button-row">
            <button type="button" class="action-button primary" data-run-single>运行单标的仲裁 Run Arbitration</button>
          </div>
        </article>

        <article class="panel-card">
          <div class="panel-head">
            <div>
              <span class="panel-kicker">组合入口 Portfolio Entry</span>
              <h3>运行组合仲裁 Run Portfolio Arbitration</h3>
            </div>
          </div>
          <form data-portfolio-form class="form-grid">
            <label><span>组合 ID Portfolio ID</span><input name="portfolio_id" value="${utils.escapeHtml(page.portfolioForm.portfolio_id)}" /></label>
            <label><span>标的 Symbol</span><input name="symbol" value="${utils.escapeHtml(page.portfolioForm.symbol)}" /></label>
            <label>
              <span>趋势方向 Trend Direction</span>
              <select name="trend_direction">
                ${["LONG", "SHORT", "FLAT"]
                  .map((option) => `<option value="${option}" ${page.portfolioForm.trend_direction === option ? "selected" : ""}>${option}</option>`)
                  .join("")}
              </select>
            </label>
            <label><span>趋势强度 Trend Strength</span><input name="trend_strength" type="number" min="0" max="1" step="0.01" value="${utils.escapeHtml(page.portfolioForm.trend_strength)}" /></label>
            <label><span>趋势置信度 Trend Confidence</span><input name="trend_confidence" type="number" min="0" max="1" step="0.01" value="${utils.escapeHtml(page.portfolioForm.trend_confidence)}" /></label>
            <label><span>趋势权重 Trend Weight</span><input name="trend_weight" type="number" min="0" max="1" step="0.01" value="${utils.escapeHtml(page.portfolioForm.trend_weight)}" /></label>
            <label>
              <span>均值方向 Mean Direction</span>
              <select name="mean_direction">
                ${["LONG", "SHORT", "FLAT"]
                  .map((option) => `<option value="${option}" ${page.portfolioForm.mean_direction === option ? "selected" : ""}>${option}</option>`)
                  .join("")}
              </select>
            </label>
            <label><span>均值强度 Mean Strength</span><input name="mean_strength" type="number" min="0" max="1" step="0.01" value="${utils.escapeHtml(page.portfolioForm.mean_strength)}" /></label>
            <label><span>均值置信度 Mean Confidence</span><input name="mean_confidence" type="number" min="0" max="1" step="0.01" value="${utils.escapeHtml(page.portfolioForm.mean_confidence)}" /></label>
            <label><span>均值权重 Mean Weight</span><input name="mean_weight" type="number" min="0" max="1" step="0.01" value="${utils.escapeHtml(page.portfolioForm.mean_weight)}" /></label>
          </form>
          <div class="button-row">
            <button type="button" class="action-button primary" data-run-portfolio>运行组合仲裁 Run Portfolio Arbitration</button>
          </div>
        </article>
      </section>

      <section class="panel-grid two-up">
        <article class="panel-card">
          <div class="panel-head">
            <div>
              <span class="panel-kicker">单标的结果 Single Result</span>
              <h3>决策卡片 Decision Card</h3>
            </div>
          </div>
          ${renderDecisionSummary(page.singleResult, utils, "提交单标的表单后，这里会显示真实仲裁结果。 / Submit the single-symbol form to see a live arbitration result.")}
        </article>

        <article class="panel-card">
          <div class="panel-head">
            <div>
              <span class="panel-kicker">组合结果 Portfolio Result</span>
              <h3>决策卡片 Decision Card</h3>
            </div>
          </div>
          ${renderDecisionSummary(page.portfolioResult, utils, "提交组合表单后，这里会显示返回的组合仲裁结果。 / Submit the portfolio form to see the returned portfolio arbitration decision.")}
        </article>
      </section>

      <section class="panel-card">
        <div class="panel-head">
          <div>
            <span class="panel-kicker">规则与理由 Rules and Rationale</span>
            <h3>单标的理由明细 Single-symbol Rationale Rows</h3>
          </div>
        </div>
        ${utils.renderTable(
          [
            { key: "signal_name", label: "信号 Signal" },
            { key: "direction", label: "方向 Direction" },
            {
              key: "confidence",
              label: "置信度 Confidence",
              render: (row) => utils.formatNumber(row.confidence),
            },
            {
              key: "contribution",
              label: "贡献值 Contribution",
              render: (row) => utils.formatNumber(row.contribution),
            },
            {
              key: "rule_adjustments",
              label: "规则调整 Adjustments",
              render: (row) => utils.escapeHtml((row.rule_adjustments || []).join(", ") || "none"),
            },
          ],
          rationale,
          { emptyText: "单标的仲裁完成后，这里会显示理由行。 / Rationale rows appear after a single-symbol arbitration run." },
        )}
      </section>

      ${utils.renderJsonPanel("单标的仲裁原始响应 Single Arbitration Raw Response", page.singleResult)}
      ${utils.renderJsonPanel("组合仲裁原始响应 Portfolio Arbitration Raw Response", page.portfolioResult)}
    `;
  },
  mount(ctx, page, root) {
    const syncSingle = () => {
      const data = new FormData(root.querySelector("[data-single-form]"));
      Object.keys(page.singleForm).forEach((key) => {
        page.singleForm[key] = data.get(key) ?? page.singleForm[key];
      });
      ["confidence", "strength"].forEach((key) => {
        page.singleForm[key] = Number(page.singleForm[key]);
      });
    };

    const syncPortfolio = () => {
      const data = new FormData(root.querySelector("[data-portfolio-form]"));
      Object.keys(page.portfolioForm).forEach((key) => {
        page.portfolioForm[key] = data.get(key) ?? page.portfolioForm[key];
      });
      ["trend_strength", "trend_confidence", "trend_weight", "mean_strength", "mean_confidence", "mean_weight"].forEach((key) => {
        page.portfolioForm[key] = Number(page.portfolioForm[key]);
      });
    };

    root.querySelector("[data-single-form]")?.addEventListener("input", syncSingle);
    root.querySelector("[data-single-form]")?.addEventListener("change", syncSingle);
    root.querySelector("[data-portfolio-form]")?.addEventListener("input", syncPortfolio);
    root.querySelector("[data-portfolio-form]")?.addEventListener("change", syncPortfolio);

    root.querySelector("[data-run-single]")?.addEventListener("click", async () => {
      syncSingle();
      ctx.setLoading(true);
      ctx.setPageStatus(this.id, {
        tone: "loading",
        title: "运行单标的仲裁 Running Arbitration",
        message: "正在提交单标的仲裁请求。",
        detail: "POST /api/v1/arbitration/run",
      });
      const result = await ctx.api.post("/api/v1/arbitration/run", page.singleForm);
      page.singleResult = result.data;
      if (result.ok) {
        ctx.setPageStatus(this.id, {
          tone: "success",
          title: "单标的仲裁完成 Single-symbol Arbitration Completed",
          message: `当前后端已返回决策 ${result.data?.bias || "n/a"}。`,
          detail: result.data?.decision_id || "decision id unavailable",
        });
        ctx.toast("单标的仲裁完成 / Single-symbol arbitration completed.", "success");
      } else {
        ctx.setPageStatus(this.id, {
          tone: "error",
          title: "单标的仲裁失败 Single-symbol Arbitration Failed",
          message: result.error?.message || "后端返回错误。",
          detail: `${result.status || "network"} / ${result.url}`,
        });
        ctx.toast("单标的仲裁失败 / Single-symbol arbitration failed.", "error");
      }
      ctx.setLoading(false);
      ctx.rerender();
    });

    root.querySelector("[data-run-portfolio]")?.addEventListener("click", async () => {
      syncPortfolio();
      ctx.setLoading(true);
      ctx.setPageStatus(this.id, {
        tone: "loading",
        title: "运行组合仲裁 Running Portfolio Arbitration",
        message: "正在提交组合仲裁请求。",
        detail: "POST /api/v1/arbitration/run-portfolio",
      });
      const result = await ctx.api.post("/api/v1/arbitration/run-portfolio", {
        portfolio_id: page.portfolioForm.portfolio_id,
        symbol: page.portfolioForm.symbol,
        proposals: [
          {
            proposal_id: "trend-proposal",
            strategy_id: "trend",
            aggregate_direction: page.portfolioForm.trend_direction,
            aggregate_strength: page.portfolioForm.trend_strength,
            aggregate_confidence: page.portfolioForm.trend_confidence,
            portfolio_weight: page.portfolioForm.trend_weight,
            bundle_count: 1,
          },
          {
            proposal_id: "mean-reversion-proposal",
            strategy_id: "mean_reversion",
            aggregate_direction: page.portfolioForm.mean_direction,
            aggregate_strength: page.portfolioForm.mean_strength,
            aggregate_confidence: page.portfolioForm.mean_confidence,
            portfolio_weight: page.portfolioForm.mean_weight,
            bundle_count: 1,
          },
        ],
      });
      page.portfolioResult = result.data;
      if (result.ok) {
        ctx.setPageStatus(this.id, {
          tone: "success",
          title: "组合仲裁完成 Portfolio Arbitration Completed",
          message: `组合仲裁已成功返回 ${result.data?.bias || "n/a"}。`,
          detail: result.data?.decision_id || "decision id unavailable",
        });
        ctx.toast("组合仲裁完成 / Portfolio arbitration completed.", "success");
      } else {
        ctx.setPageStatus(this.id, {
          tone: "error",
          title: "组合仲裁失败 Portfolio Arbitration Failed",
          message: result.error?.message || "后端返回错误。",
          detail: `${result.status || "network"} / ${result.url}`,
        });
        ctx.toast("组合仲裁失败 / Portfolio arbitration failed.", "error");
      }
      ctx.setLoading(false);
      ctx.rerender();
    });
  },
};
